### DISCUSSION TITLE
[Solved] XAML Extension for GravatarImageSource

### DISCUSSION THREAD

**Discussion by egvijayanand** [CommunityToolkit/Maui#679] [New Feature Discussions]
Though it might be a simple extension, it simplifies the XAML to a greater extent and improves readability.

Before:

```xml
<Image>
  <Image.Source>
    <mct:GravatarImageSource Email="user@example.com" />
  </Image.Source>
</Image>
```

After:

```xml
<Image Source="{mct:GravatarImage 'user@example.com'}" />
```
Implementation:

```csharp
namespace CommunityToolkit.Maui.Extensions
{
    [ContentProperty(nameof(Email))]
    public class GravatarImageExtension : GravatarImageSource, IMarkupExtension<GravatarImageSource>, IMarkupExtension
    {
        public GravatarImageSource ProvideValue(IServiceProvider serviceProvider)
        {
            ArgumentNullException.ThrowIfNull(serviceProvider);

            if (BindingContext == null && !IsSet(BindingContextProperty))
            {
                if (serviceProvider.GetService(typeof(IProvideValueTarget)) is IProvideValueTarget val)
                {
                    SetBinding(BindingContextProperty, new Binding(nameof(BindingContext), BindingMode.Default, null, null, null, val.TargetObject));
                }
            }

            return this;
        }

        object IMarkupExtension.ProvideValue(IServiceProvider serviceProvider)
            => ProvideValue(serviceProvider);
    }
}
```

**Comment by TheCodeTraveler** [comment]
@bijington @GeorgeLeithead - what do you think?

**Comment by pictos** [comment]
Hey, @egvijayanand, thanks for your proposal.
This isn't needed, you can already do that in .NET MAUI, as you can see in the screenshot below.

Let's say that .NET MAUI has a secret gem that allows you to describe any control using markups, so you don't need to create markup for these types, for example, let's say that you want to do a rounded border, you can do that in two ways:

```xml
			<Border
				StrokeThickness="4">
				<Border.StrokeShape>
					<RoundRectangle CornerRadius="40,0,0,40" />
				</Border.StrokeShape>
				<Label  Text="Hello, World!" />
			</Border>
```
or 

```xml
<Border
	StrokeShape="{RoundRectangle CornerRadius='40,0,0,40'}"
	StrokeThickness="4">
	<Label
		FontSize="32"
		HorizontalOptions="Center"
		SemanticProperties.HeadingLevel="Level1"
		Text="Hello, World!" />
</Border>
```

![image](https://user-images.githubusercontent.com/20712372/194154690-5219ece8-f113-4b53-af41-f5d40cd323fb.png)

**Reply by TheCodeTraveler** [reply]
Oh cool! Could you update the docs to add this info, @pictos? https://learn.microsoft.com/dotnet/communitytoolkit/maui/imagesources/gravatarimagesource

Right now, the docs only specify how Vijay demonstrated in his `Before` example.

**Reply by egvijayanand** [reply]
> Let's say that .NET MAUI has a secret gem that allows you to describe any control using markups, so you don't need to create markup for these types, for example, let's say that you want to do a rounded border, you can do that in two ways:

Pro Tip, have used the Border the way it is described in the sample and thought a XAML extension is defined in the framework itself. But this is definitely a much-needed feature as it'll save so many keystrokes and avoid unnecessary nesting.

**Comment by egvijayanand** [comment]
@brminnick Since the special feature of .NET MAUI makes it super easy to define the Source without any need for an explicit extension, this discussion can be marked as resolved/answered.

**Comment by CliffAgius** [comment]
Closed and answered
